v2/dtos/requests/device_test.go

```diff
@@ -0,0 +1,159 @@
+	noAutoEventFrequency := testAddDevice
+	noAutoEventFrequency.Device.AutoEvents = []dtos.AutoEvent{
+		{Resource: "TestDevice", OnChange: true},
+	}
+	noAutoEventResource := testAddDevice
+	noAutoEventResource.Device.AutoEvents = []dtos.AutoEvent{
+		{Frequency: "300ms", OnChange: true},
+	}
+	tests := []struct {
+		name        string
+		Device      AddDeviceRequest
+		expectError bool
+	}{
+		{"valid AddDeviceRequest", valid, false},
+		{"invalid AddDeviceRequest, invalid autoEvent frequency", invalidFrequency, true},
+		{"invalid AddDeviceRequest, no Request Id", noReID, true},
+		{"invalid AddDeviceRequest, no DeviceName", noDeviceName, true},
+		{"invalid AddDeviceRequest, no ServiceName", noServiceName, true},
+		{"invalid AddDeviceRequest, no ProfileName", noProfileName, true},
+		{"invalid AddDeviceRequest, no Protocols", noProtocols, true},
+		{"invalid AddDeviceRequest, no AutoEvent frequency", noAutoEventFrequency, true},
+		{"invalid AddDeviceRequest, no AutoEvent resource", noAutoEventResource, true},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := tt.Device.Validate()
+			assert.Equal(t, tt.expectError, err != nil, "Unexpected addDeviceRequest validation result.", err)
+		})
+	}
+}
+
+func TestAddDevice_UnmarshalJSON(t *testing.T) {
+	valid := testAddDevice
+	resultTestBytes, _ := json.Marshal(testAddDevice)
+	type args struct {
+		data []byte
+	}
+	tests := []struct {
+		name      string
+		addDevice AddDeviceRequest
+		args      args
+		wantErr   bool
+	}{
+		{"unmarshal AddDeviceRequest with success", valid, args{resultTestBytes}, false},
+		{"unmarshal invalid AddDeviceRequest, empty data", AddDeviceRequest{}, args{[]byte{}}, true},
+		{"unmarshal invalid AddDeviceRequest, string data", AddDeviceRequest{}, args{[]byte("Invalid AddDeviceRequest")}, true},
+	}
+	fmt.Println(string(resultTestBytes))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var expected = tt.addDevice
+			err := tt.addDevice.UnmarshalJSON(tt.args.data)
+			if tt.wantErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, expected, tt.addDevice, "Unmarshal did not result in expected AddDeviceRequest.")
+			}
+		})
+	}
+}
+
+func Test_AddDeviceReqToDeviceModels(t *testing.T) {
+	valid := []AddDeviceRequest{testAddDevice}
+	expectedAutoEvents := dtos.AutoEventDTOsToModels(autoEvents)
+	expectedProtocols := dtos.ProtocolDTOsToModels(protocols)
+	expectedDeviceModel := []models.Device{{
+		Name:           TestDeviceName,
+		ServiceName:    TestDeviceServiceName,
+		ProfileName:    TestDeviceProfileName,
+		AdminState:     models.Locked,
+		OperatingState: models.Enabled,
+		Labels:         testDeviceLabels,
+		Location:       testDeviceLocation,
+		AutoEvents:     expectedAutoEvents,
+		Protocols:      expectedProtocols,
+	}}
+	tests := []struct {
+		name       string
+		addDevices []AddDeviceRequest
+	}{
+		{"valid AddDeviceRequest", valid},
```

Comment:
Do you plan to add invalid test case? If, not change to non-table driven to reduce complexity.

Reply:
Changed to non-table driven testing.